src/memory.rs

```diff
@@ -133,69 +154,79 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
         }
     }
 
-    pub fn allocate(&mut self, size: usize) -> EvalResult<'tcx, Pointer> {
+    pub fn allocate(&mut self, size: usize, align: usize) -> EvalResult<'tcx, Pointer> {
         if size == 0 {
             return Ok(Pointer::zst_ptr());
         }
+        // make sure we can offset the result pointer by the worst possible alignment
+        // this allows cheaply checking for alignment directly in the pointer
+        let least_aligned_size = size + align;
```

Comment:
I don't really understand this, but it seems fishy to me. The first thing I noticed is that a simple array of chars (say, size 10, align 1) will have an extra byte for no reason. Can we accomplish this without wasting any bytes?